keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
+        // +optional
+        Used ResourceList
++       // QoSResources contains the enforced set of available class resources.
++       // +featureGate=QoSResources
++       // +optional
++       QoSResources QoSResourceQuota
+ }
+
+
+```
+## Proposal
+
+This section currently covers [implementation phase 1](#phase-1) (see
+[implementation phases](#implementation-phases) for an outline of the complete
+implementation).
+
+We extend the CRI protocol to contain information about the QoS-class
+resource assignment of containers and pods. Resource assignment requests will
+be simple key-value pairs (*resource-type=class-name*)
+
+We also extend the CRI protocol to support updates of QoS-class resource
+assignment of running containers. We recognize that currently container
+runtimes lack the capability to update either of the two types of QoS-class
+resources we have identified (RDT and blockio). However, there is no technical
+limitation in that and we are planning to implement update support for them
+in the future.
+
+We also extend the Kubernetes API to support the assignment of QoS-class
+resources to applications via PodSpec, and to users and the kube-scheduler to
+see availability of QoS-class resources on a per-node basis via NodeStatus.
+
+Container runtime is expected to be aware of all resource types and the classes
+within. The CRI protocol is extended to be able to communicate the available
+QoS-class resources from the runtime to the client. This information includes:
+- Available QoS-class resource types.
+- Available classes within each resource type.
+- Whether the resource type is immutable or if it supports in-place updates.
+  In-place updates of resoures might not be possible because of runtime
+  limitations or the underlying technology, for example.
+
+QoS-class resources may be bounded in the way that the number of applications
+that can be assigned to a specific class (of one QoS-class resource) on a node
+can be limited. This limit is configuratble on a per-class (and per-node)
+basis. This can be used to e.g. limit access to a high-tier class.
+
+Pod-level and container-level QoS-class resources are completely independent
+resource types. E.g. specifying something in the pod-level request does not
+mean specifying a pod-level default for all containers of the pod.
+
+Currently we identify two types of container-level QoS-class resources (RDT and
+blockio) but the API changes will be generic so that it will serve other
+similar resources in the future. Currently there are no immediately enabled
+pod-level QoS-class resources but we see usage scenarios for those in the
+future (communicating the pod QoS class to the runtime and enabling pod-level
+cgroup controls for blockio).
+
+We introduce a feature gate that enables Kubernetes components (kubelet,
+kube-apiserver, kube-scheduler) to support QoS-class resources.
+
+### User Stories (Optional)
```

Comment:
can you please think of any ways this can be abused for non-resource related features? With the cases of possible abuse, will there be alternative approaches k8s already supports today that would work better?

Examples I can think of are: 
- using QoS class instead of node labels to indicate some features
- using QoS class like an alterative for pod and node affinity
- replacement of runtimehandler, priority class (better limiting of priority class pods)?
- if there will be external call during pod admission - can this be abused as generic pod admission plugin system?

Reply:
> can you please think of any ways this can be abused for non-resource related features? With the cases of possible abuse, will there be alternative approaches k8s already supports today that would work better?

This is a good commit. Overall when I now looked at the user stories they are way too brief and narrow in perspective (the KEP has evolved quite a bit but the user stories haven't been touched much).

I think the QoS resources mechanism could be "abused" for many different purposes as it's so opaque to kubernetes. Sort of "extended resources with a class identifier instead of numerical capacity". So it could be "abused" in many scenarios where you need to assign key=value pairs (where value is an identifier string out of a known set).

At a quick thought, I think all of the three below would be at least somewhat applicable. 

> * using QoS class instead of node labels to indicate some features

Possibly, in some scenarios. However with labels you can be much more flexible with the different node affinity options (affinity/anti-affinity, required vs preferrred, complex match expressions etc

> * using QoS class like an alterative for pod and node affinity

Again, possibly in some scenarios but I think QoS resources (in the proposed form) would be quite limited wrt. existing options.

> * replacement of runtimehandler, priority class (better limiting of priority class pods)?

This is actually a quick thought I also had at some point but I haven't really done the homework to think this properly through. So maybe (some other people might have better idea about this)

> * if there will be external call during pod admission - can this be abused as generic pod admission plugin system?

We were not thinking about a separate pod admission call (e.g. in CRI) so I don't think this is currently applicable.

Reply:
changed wording "abused" -> "used"